src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -4853,6 +4853,80 @@ TEST_CASE("TransposedConvolutionLayerPaddingTest", "[ANNLayerTest]")
 //   REQUIRE(output.n_elem == 4);
 // }
 
+/**
+ * Simple test for Residual layer.
+ */
+TEST_CASE("ResidualTestCase", "[ANNLayerTest]")
+{
+  // For rectangular input to pooling layers.
+  arma::mat input = arma::mat(28, 1);
+  input.zeros();
+  input(0) = input(16) = 1;
+  input(1) = input(17) = 2;
+  input(2) = input(18) = 3;
+  input(3) = input(19) = 4;
+  input(4) = input(20) = 5;
+  input(5) = input(23) = 6;
+  input(6) = input(24) = 7;
+  input(14) = input(25) = 8;
+  input(15) = input(26) = 9;
+
+  Residual module1;
+  module1.Add<MeanPooling>(2, 2, 2, 2, false);
+  module1.Add<MeanPooling>(2, 2, 2, 2, false);
+
+  Residual module2;
+  module2.Add<MeanPooling>(2, 2, 2, 2, true);
+  module2.Add<MeanPooling>(2, 2, 2, 2, true);
```

Comment:
It might be clearer to use a network that is simpler to reason about.  For instance, instead of a MeanPooling layer, maybe it makes more sense to have a Linear layer with weights equal to 2, and a Linear layer with weights equal to -1, and then ensure that the output is identical (for forward and backward passes) to a single Linear layer with weights equal to 1, and also ensure that the 'merged' gradient of the two linear layers is also identical to the gradient of the single Linear layer.

I think a lot of the other tests for layers in the past have not been very comprehensive, so it is not a bad idea to go a bit above and beyond here. :)

Reply:
Can you give me example please?

Reply:
Sorry for the slow response! I somehow did not see the email notification for this.  What I mean is to use two networks like this:

```
// A bit tedious, we have to manage the memory ourselves...
Residual r;
r.Add<Linear>(5);
r.Add<Linear>(5);
r.InputDimensions() = std::vector<size_t>({ 5 });
r.ComputeOutputDimensions();
arma::mat rParams(r.WeightSize(), 1);
r.SetWeights((double*) rParams.memptr());
r.Network()[0].Parameters().fill(2.0);
r.Network()[1].Parameters().fill(-1.0);

Linear l(5);
l.InputDimensions() = std::vector<size_t>({ 5 });
l.ComputeOutputDimensions();
arma::mat lParams(l.WeightSize(), 1);
l.SetWeights((double*) lParams.memptr());
l.Parameters().fill(1.0);
```

Now, if I pass data through `r` and `l`, the forward and backward passes should produce the same results.  For the gradient the check is a little more tricky---you'll have to add the gradients from both residual layers together, but they should be equal to the gradient from `l`.

Or, if you are asking for some other test ideas, here are a few:

 * Make a Residual layer with only one child layer (here you can use a different layer type with random weights), and ensure that the output for forward/backward/gradient is exactly the same as if you have the child layer on its own.
 * Make sure that Residual works with batch size greater than one (to do this, just pass an input with multiple columns).
 * If you are feeling tricky, make sure that Residual acts as expected when there are no child layers at all.
 * If you create a Residual layer with child layers that do not have the same output dimensions, ensure that an exception is thrown.

Those are just a couple ideas, maybe you can come up with others?  Hopefully I answered your question well, if not, let me know and I can try again. :)

Reply:
By seeing this test case, I have a question that is it necessary to provide method `Parameters()` inside layers as I haven't given this in the AddMerge layer?

Reply:
The `AddMerge` layer itself has no parameters, so there isn't a need for the `Parameters()` function.  `Parameters()` is defined in the base `Layer` class, but it throws an exception if you try to call it for a layer that hasn't overloaded it.